Add comparison functions

cc7768 opened the pull request:
This pull request adds a new group of binary operators for comparisons by implementing `(==, <, <=, >, >=)` instead of just their element-wise counterparts.

This isn't quite ready to merge. One lingering implementation issue corresponds to the points raised in #63 with the type stability enforced by the `FixedSizeArrays` version of `map`. The new operators rely on being able to call `all(x .? y)` which won't work unless the output of `x .? y` is a boolean.

Don't know if this will be helpful, but I thought I'd throw this out for review.

cc7768 commented:
Also, still need to write tests.

SimonDanisch commented:
Thanks! Yeah, I definitely need to change the behaviour of map. Right now I'm tight with time so it won't be happening too soon! But I'm sort of on it ;)
This looks good, especially when there are tests!

cc7768 commented:
Not a problem. To make this useful sooner, I added a **temporary** hack and used `map(Bool, map(operation, x, y))` -- Admittedly not pretty, but I don't know enough to tackle the `map` problem itself.

I added a few (pretty trivial) tests. I'll add a few more as I find/make some time.

One small thing that could be made a little more informative is the error message from comparing (or applying other binary operations) FixedSizeArrays of different sizes. Currently it gives:

```
julia> Vec(1, 2, 3) < Vec(1.0, 2.0, 3.0, 4.0)
ERROR: StackOverflowError:
 in .< at /home/chase/.julia/v0.4/FixedSizeArrays/src/ops.jl:84 (repeats 257 times)
```

because it keeps trying to promote the two to the same type when it can't because of size issues. It might be nice to throw a size error instead (I don't know whether implementing this would be more trouble than its worth).

c42f commented:
I've been working on map which will solve the output type problems here, and I just noticed this PR.

I thought I'd weigh in against having the ordering operators defined at all, especially not in terms of `any()` or `all()`.  The problem is that either of these options can give apparently nonsensical results, for example, if `<=(vec1, vec2) = all(vec1 .<= vec2)`, then you can have `!(a <= b)` and `!(a >= b)` true simultaneously (take `a = Vec(1,2)` and `b = Vec(2,1)` for example).

It is true that `<=` defined as above is a [partial order](https://en.wikipedia.org/wiki/Partially_ordered_set) which is technically what's required for `Base.<`, but it's in no way the definitive partial ordering.  Arguably a better choice would be to choose something like lexicographic ordering, in which case you'd at least have a [total order](https://en.wikipedia.org/wiki/Total_order) whenever the `T` in `Vec{N,T}` is itself totally ordered.  For a total order you can have the operators `<`, `>`, `<=` and `>=` self consistently defined.

Even lexicographic ordering isn't really the definitive ordering though; there just isn't one, so I suggest we don't define it ;-)  Some related background here http://math.stackexchange.com/questions/1620004/ordering-of-complex-plane

mschauer commented:
Relevant: JuliaLang/julia#8343

cc7768 commented:
Thanks for bringing these things up. Haven't looked at (or worked on) this in awhile -- Things caught up to me and got distracted.

Good point about total vs partial order. In a lot of the work that I do, the partial order comparison I wrote down is what we typically use. I didn't think about the fact that other people possibly do different things -- This is why the review process is useful :) Makes sense not to implement operators that different people want to use differently -- I think the element wise operators do exactly what everyone suspects, but some will think `<` implements a lexicographic order while others will think it implements the partial order I had in mind (and others might have different comparisons in mind). Once the `map` problem is finished it is easy enough for me to use `all(x .< y)`.

c42f commented:
Thanks for taking the time to make a patch :-)
